include/clap/ext/thread-check.h

```diff
@@ -17,23 +17,40 @@ extern "C" {
+///    or asset loading in dedicated background threads started by the plugin.
 ///
 /// audio-thread:
-///    This thread is used for realtime audio processing. Its execution should be as deterministic
-///    as possible to meet the audio interface's deadline (can be <1ms). In other words, there is a
-///    known set of operations that should be avoided: malloc() and free(), mutexes (spin mutexes
-///    are worse), I/O, waiting, ...
-///    The audio-thread is something symbolic, there isn't one OS thread that remains the
-///    audio-thread for the plugin lifetime. As you may guess, the host is likely to have a
+///    This thread can be used for realtime audio processing. Its execution should be as deterministic
+///    as possible to meet the audio interface's deadline (can be <1ms). There are a
+///    known set of operations that should be avoided: malloc() and free(), contended locks and mutexes,
+///    I/O, waiting, and so forth.
+///
+///    The audio-thread is symbolic, there isn't one OS thread that remains the
+///    audio-thread for the plugin lifetime. A host is may opt to have a
 ///    thread pool and the plugin.process() call may be scheduled on different OS threads over time.
-///    The most important thing is that there can't be two audio-threads at the same time. All the
-///    functions marked with [audio-thread] **ARE NOT CONCURRENT**. The host may mark any OS thread,
+///    However, the host must guarantee that single plugin instance will not be two audio-threads
+///    at the same time.
+///
+///    Functions marked with [audio-thread] **ARE NOT CONCURRENT**. The host may mark any OS thread,
 ///    including the main-thread as the audio-thread, as long as it can guarantee that only one OS
-///    thread is the audio-thread at a time. The audio-thread can be seen as a concurrency guard for
-///    all functions marked with [audio-thread].
+///    thread is the audio-thread at a time in a plugin instance. The audio-thread can be seen as a
+///    concurrency guard for all functions marked with [audio-thread].
+///
+///    The real-time constraint on the [audio-thread] interacts closely with the render extension.
+///    If a plugin doesn't implement render, then that plugin must have all [audio-thread] functions
+///    meet the real time standard. If the plugin does implement render, and returns true when
+///    render mode is set to real-time or if the plugin advertises a hard realtime requirement, it
+///    must implement realtime constraints. Host provided [audio-thread] functions can be called
+///    from a plugin on the audio thread and must also therefore either meet the real-time constraint,
+///    not run plugins which advertise a hard realtime constraint via render or don't implement render,
+///    or expect potentially inconsistent rendering and performance from plugins when they re-enter the host.
+///
```

Comment:
This last long sentence is a bit hard to read. 

-  `and must also therefore` the 'also' is not needed
- `,not run plugins ...` also reads strange

I would suggest:

```
Host-provided [audio-thread] functions can be called from a plugin on the audio thread and must, therefore, either meet the real-time constraint, avoid running plugins that advertise a hard real-time constraint via render or don't implement render, or expect potentially inconsistent rendering and performance from plugins when they re-enter the host.
``` 

but IMO this sentence is still too long.

Reply:
@baconpaul ⬆️

Reply:
On this question, I prefer to let you discuss, I'm not very good at writing precise docs in correct English ;-)

Reply:
Here's a cleaner version. Original

> Host provided [audio-thread] functions can be called
> from a plugin on the audio thread and must also therefore either meet the real-time constraint,
> not run plugins which advertise a hard realtime constraint via render or don't implement render,
> or expect potentially inconsistent rendering and performance from plugins when they re-enter the host.

Updated

> Hosts also provide functions marked [audio-thread]. These can be safely called by a plugin in the audio 
> thread. Therefore hosts must either (1) implement those functions meeting the real-time constraints or
> (2) not process plugins which advertise a hard realtime constraint or don't implement the render extension. 
> Hosts which provide [audio-thread] functions outside these conditions may experience inconsistent or 
> inaccurate rendering.

thoughts?

Reply:
Sounds good to me!

Reply:
I copy the updated text then, thanks!

Reply:
Done.